apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSource.java

```diff
@@ -0,0 +1,59 @@
+package gov.cms.bfd.pipeline.rda.grpc.server;
+
+import java.util.NoSuchElementException;
+import java.util.function.Predicate;
+
+/**
+ * Wrapper around a {@link MessageSource} to allow the values to be filtered using some condition.
+ * Generally the filter will be a minimum sequence number.
+ *
+ * @param <T> type of objects being returned by the source.
+ */
+public class FilteredMessageSource<T> implements MessageSource<T> {
+  private final MessageSource<T> source;
+  private final Predicate<T> filter;
+  private T nextValue;
+
+  /**
+   * Wrap the specified source and filter its objects using the supplied predicate. The predicate
+   * must return true for objects that are valid and false for objects that are invalid.
+   *
+   * @param source real source of objects
+   * @param filter function returning true for objects to keep or false for objects to skip
+   */
+  public FilteredMessageSource(MessageSource<T> source, Predicate<T> filter) {
+    this.source = source;
+    this.filter = filter;
+  }
+
+  @Override
+  public boolean hasNext() throws Exception {
+    // allow for cases of hasNext() being called multiple times in a row
+    if (nextValue != null) {
+      return true;
+    }
+
+    while (source.hasNext()) {
+      nextValue = source.next();
+      if (filter.test(nextValue)) {
+        return true;
+      }
+    }
+    return false;
```

Comment:
I was going to comment on sequence number order here, as it could affect the way the filter is being used here, but it's probably reasonable to assume sequence numbers will be in order.

Reply:
I agree with your logic and the thought that led to the question. This class is independent of sequence numbers so the order doesn't come into play here.  Generally the message sources needs to be provided with data in proper order.  For example JsonMessageSource doesn't sort the file because that could be very slow.